File: Tree/101. Symmetric Tree/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# First Try
# class Solution:
#     def isSymmetric(self, root):
#         if not root: return True

#         def dfs(left, right):
#             if not left and not right: return True
#             if not left or not right: return False
#             return all((
#                 left.val == right.val,
#                 dfs(left.left, right.right),
#                 dfs(left.right, right.left)
#             ))

#         return dfs(root.left, root.right)

# This solution is more clear
class Solution:
    def isSymmetric(self, root):
        def dfs(left, right):
            if not left and not right: return True
            if not left or not right: return False
            return all((
                left.val == right.val,
                dfs(left.left, right.right),
                dfs(left.right, right.left)
            ))

        return dfs(root, root)

File: Tree/101. Symmetric Tree/go/main.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isSymmetric(root *TreeNode) bool {
	if root == nil {
		return true
	}

	return areSymmetric(root.Left, root.Right)
}

func areSymmetric(left *TreeNode, right *TreeNode) bool {
	if left == nil && right == nil {
		return true
	}

	if (left == nil && right != nil) ||
		(left != nil && right == nil) ||
		(left.Val != right.Val) {
		return false
	}

	return areSymmetric(left.Left, right.Right) && areSymmetric(left.Right, right.Left)
}
